language/diem-vm/mvhashmap/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::OnceCell;
+use std::{
+    cmp::{max, PartialOrd},
+    collections::{btree_map::BTreeMap, HashMap},
+    hash::Hash,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+/// A structure that holds placeholders for each write to the database
+//
+//  The structure is created by one thread creating the scheduling, and
+//  at that point it is used as a &mut by that single thread.
+//
+//  Then it is passed to all threads executing as a shared reference. At
+//  this point only a single thread must write to any entry, and others
+//  can read from it. Only entries are mutated using interior mutability,
+//  but no entries can be added or deleted.
+//
+
+pub type Version = usize;
+
+const FLAG_UNASSIGNED: usize = 0;
+const FLAG_DONE: usize = 2;
+const FLAG_SKIP: usize = 3;
+
+pub struct MVHashMap<K, V> {
+    data: HashMap<K, BTreeMap<Version, WriteVersionValue<V>>>,
+}
+
+#[cfg_attr(any(target_arch = "x86_64"), repr(align(128)))]
+pub(crate) struct WriteVersionValue<V> {
+    flag: AtomicUsize,
+    data: OnceCell<Option<V>>,
```

Comment:
I still find the following design without the flag indicator more elegant (design and probably code too), but checking skip would be slightly less efficient (may not matter overall):


OnceCell .get returns None -> UNASSIGNED

otherwise it's OnceCell< enum among DELETE / SKIP / VALUE<V> >

This to me describes what's going on a bit better, and should simplify the checks (no separate flag) / cases?